### PR TITLE
Added Farsi

### PR DESCRIPTION
**Pull request by mdj117**
Well that took a while, added almost 100% of the translated version of Terasology in Farsi
The flag is named flag_fa.png and the file is menu_fa.lang

### CODE REVIEW

**Comment by GooeyHub** [comment]
Can one of the admins please verify this patch?

File: engine/src/main/resources/assets/i18n/menu_fa.lang [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+    "binding-crouch": "قوز کردن / پایین رفتن",
+    "binding-crouch-mode": "حالت نشسته",
+    "binding-decrease-view-distance": "کاهش فاصله ی دید",
+    "binding-forwards": "رو به جلو",
+    "binding-frob": "هدف استفاده",
+    "binding-hide-hud": "اچ یو دی مخفی",
+    "binding-increase-view-distance": "افزایش فاصله ی دید",
+    "binding-jump": "پرش",
+    "binding-left": "گردش از چپ",
+    "binding-nui-editor": "NUI Editor",
```
Comment (rzats) [inline engine/src/main/resources/assets/i18n/menu_fa.lang]: English string here - can you wrap this up just so the language is 100% finished? [1]

File: engine/src/main/resources/assets/i18n/menu_fa.lang [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+    "binding-nui-skin-editor": "ان. یو. آی ادیتور اسکین",
+    "binding-pause": "منوی انتخاب در حین بازی",
+    "binding-right": "گردش از راست",
+    "binding-show-online-players": "نمایش بازیکنان انلاین",
+    "binding-toggle-speed-permanently": "دکمه ی سرعت دایمی",
+    "binding-toggle-speed-temporarily": "دکمه ی سرعت موقتی",
+    "binding-use-item": "استفاده از آیتم نگه داشته شده",
+    "bloom-effect": "بلووم                        ⁠⁠",
```
Comment (rzats) [inline engine/src/main/resources/assets/i18n/menu_fa.lang]: Some odd padding at the end here; probably unintentional.

File: engine/src/main/resources/assets/i18n/menu_fa.lang [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+    "camera-setting-normal": "تنظیمات معمولی دوربین",
+    "camera-setting-smooth": "نرم",
+    "category-behavior": "دسته بندی عمومی رفتار",
+    "category-general": "دسته بندی عمومی رفتار",
+    "category-interaction": "اثر متقابل",
+    "category-inventory": "دسته بندی -بازبینی",
+    "category-mouse": "موش واره",
+    "category-movement": "حرکت",
+    "category-nui": "NUI",
+    "change-keybind-popup-message": "is already bound to an action. Rebind anyway?",
```
Comment (rzats) [inline engine/src/main/resources/assets/i18n/menu_fa.lang]: This is untranslated - can you wrap this up just so the language is 100% finished? [2]

File: engine/src/main/resources/assets/i18n/menu_fa.lang [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+    "change-keybind-popup-title": "کلید هم اکنون مهیا است",
+    "clamp-lighting": "مهار نور",
+    "cloud-shadows": "سایه های ابر",
+    "connecting-to": "Connecting to",
```
Comment (rzats) [inline engine/src/main/resources/assets/i18n/menu_fa.lang]: This is untranslated - can you wrap this up just so the language is 100% finished? [3]

File: engine/src/main/resources/assets/i18n/menu_fa.lang [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+    "dev-tools-title": "ابزار توسعه دهنده",
+    "dialog-cancel": "لغو کردن",
+    "dialog-ok": "درست است",
+    "disable-all-modules": "از کار انداختن همه",
+    "download-module": "دانلود",
+    "downloading-server-list": "دانلود لیست کارساز..",
+    "drop-item": "قطع مطلب",
+    "edit-server": "ویرایش",
+    "edit-server-title": "اضافه/ویرایش  کارساز (سرور)",
+    "enter-username-message": "Hello and welcome to Terasology's multiplayer!\nWould you like to use a non-default username?",
```
Comment (rzats) [inline engine/src/main/resources/assets/i18n/menu_fa.lang]: This is untranslated - can you wrap this up just so the language is 100% finished? [4]

File: engine/src/main/resources/assets/i18n/menu_fa.lang [hunk 6]
```diff
@@ -0,0 +1,244 @@
...
+    "failed-to-join": "عدم موفقیت در پیوستن",
+    "flickering-lights": "نورهای لرزان",
+    "fog": "مه",
+    "fov-amount": "میزان میدان دید",
+    "framerate-limit": "محدودیت سازه",
+    "game-name": "اسم بازی",
+    "game-settings": "تنظیمات",
+    "game-worlds": "دنیاها",
+    "gameplay": "محتویات نمایش بازی",
+    "gameplay-info": "can click 'modules' to customize further",
```
Comment (rzats) [inline engine/src/main/resources/assets/i18n/menu_fa.lang]: This is untranslated - can you wrap this up just so the language is 100% finished? [5]

File: engine/src/main/resources/assets/i18n/menu_fa.lang [hunk 7]
```diff
@@ -0,0 +1,244 @@
...
+    "disable-all-modules": "از کار انداختن همه",
+    "download-module": "دانلود",
+    "downloading-server-list": "دانلود لیست کارساز..",
+    "drop-item": "قطع مطلب",
+    "edit-server": "ویرایش",
+    "edit-server-title": "اضافه/ویرایش  کارساز (سرور)",
+    "enter-username-message": "Hello and welcome to Terasology's multiplayer!\nWould you like to use a non-default username?",
+    "enum-editor-prompt": "انتخاب مقدار برای منحنی",
+    "enum-editor-title": "ویرایشگر منحنی داده ها",
+    "error-downloading-server-list": "اشتباه در دانلود فهرست کترساز (سرور)",
```
Comment (rzats) [inline engine/src/main/resources/assets/i18n/menu_fa.lang]: Two closing parentheses here.

File: engine/src/main/resources/assets/i18n/menu_fa.lang [hunk 8]
```diff
@@ -0,0 +1,244 @@
...
+    "join-server-refresh": "تازه کردن",
+    "join-server-action": "پیوستن",
+    "join-server-requested": "تقاضا شده",
+    "join-server-title": "اتصال به سرور",
+    "light-shafts": "پرتوهای نور",
+    "listed-servers": "لیست شده",
+    "load-game": "ذخیره گذاری",
+    "menu-animations": "فهرست فیلمهای نقاشی شده",
+    "missing-name-message": "شما باید اسم را وارد کنید",
+    "module-dependencies-empty": "This module has no dependencies",
```
Comment (rzats) [inline engine/src/main/resources/assets/i18n/menu_fa.lang]: This is untranslated - can you wrap this up just so the language is 100% finished? [6]

File: engine/src/main/resources/assets/i18n/menu_fa.lang [hunk 9]
```diff
@@ -0,0 +1,244 @@
...
+    "module-dependencies-exist": "توابع طرحها",
+    "module-filter": "فیلتر",
+    "module-list": "طرح ها",
+    "module-name": "اسم",
+    "module-status": "وضعیت",
+    "module-status-activated": "فعال شده",
+    "module-status-activegameplay": "فعال کردن گیم پلی",
+    "module-status-available": "در دسترس",
+    "module-status-dependency": "تبعیت",
+    "module-status-error": "Incompatible or unresolved dependencies",
```
Comment (rzats) [inline engine/src/main/resources/assets/i18n/menu_fa.lang]: This is untranslated - can you wrap this up just so the language is 100% finished? [7]

File: engine/src/main/resources/assets/i18n/menu_fa.lang [hunk 10]
```diff
@@ -0,0 +1,244 @@
...
+    "parsing-content": "شکافتن محتویات ..",
+    "player-color": "رنگ بازیگر",
+    "player-name": "اسم بازیگر",
+    "player-height": "قد بازیگر",
+    "player-eye-height": "ارتفاع دید بازیگر",
+    "player-settings": "بازیگر",
+    "player-settings-title": "تنظیمات بازیگر",
+    "please-wait": "لطفا صبر کنید",
+    "preview-world-title": "پیش نمایش کلمه",
+    "preview-zoom-factor": "Zoom",
```
Comment (rzats) [inline engine/src/main/resources/assets/i18n/menu_fa.lang]: This is untranslated - can you wrap this up just so the language is 100% finished? [8]

File: engine/src/main/resources/assets/i18n/menu_fa.lang [hunk 11]
```diff
@@ -0,0 +1,244 @@
...
+    "view-distance-blind": "بسته شدن قانونی",
+    "view-distance-extreme": "حداکثر",
+    "view-distance-far": "دور",
+    "view-distance-mega": "مگا",
+    "view-distance-moderate": "میانه",
+    "view-distance-near": "نزدیک",
+    "view-distance-ultra": "فرا",
+    "vignette": "تصویر",
+    "volumetric-fog": "مه حجمی",
+    "warn-modules": "اخطار: Enabling extra modules, especially all or those not in the stable lineup can cause game crashes or broken worlds.\nPlease handle with care.",
```
Comment (rzats) [inline engine/src/main/resources/assets/i18n/menu_fa.lang]: This is untranslated - can you wrap this up just so the language is 100% finished? [9]

File: engine/src/main/resources/assets/i18n/menu_fa.lang [hunk 12]
```diff
@@ -0,0 +1,244 @@
...
+    "delete-game": "حذف کردن",
+    "dev-tools-bt-editor": "نمایشگر درختی ویرایش رفتاری",
+    "dev-tools-nui-editor": "ویرایشگر صفحه ان. یو. ال",
+    "dev-tools-skin-editor": "ویرایشگر اسکین ان.یو. ال",
+    "dev-tools-title": "ابزار توسعه دهنده",
+    "dialog-cancel": "لغو کردن",
+    "dialog-ok": "درست است",
+    "disable-all-modules": "از کار انداختن همه",
+    "download-module": "دانلود",
+    "downloading-server-list": "دانلود لیست کارساز..",
```
Comment (rzats) [inline engine/src/main/resources/assets/i18n/menu_fa.lang]: Should probably be three dots.

**Comment by rzats** [comment]
@GooeyHub: ok to test

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

File: engine/src/main/resources/assets/i18n/menu_fa.lang (hunk 1 above)
Comment (mdj117) [inline engine/src/main/resources/assets/i18n/menu_fa.lang]: couldn't figure out a word for NUI editor :(

File: engine/src/main/resources/assets/i18n/menu_fa.lang (hunk 3 above)
Comment (mdj117) [inline engine/src/main/resources/assets/i18n/menu_fa.lang]: no word for re bind that fits this description :(

**Comment by msteiger** [comment]
Thanks @mdj117 for the translation. I assume that we need to add the corresponding character glyphs first. Can you specify the unicode range that is needed to cover the (basic) language?

Also, afaik, Farsi is written right-to-left, so we might need to support that first. Is that correct? What changes are needed for RTL languages?

**Comment by mdj117** [comment]
I don't really think any changes are necessary, I mean I can read it. Although we may have some problem with other RTL languages.
The range used was 0600–06FF, which covers basic Arabic scripts.

**Comment by msteiger** [comment]
I've added Arabic supplements as well (only 48 glyphs) to fully cover the Naskh script. It looks OK, but too small. Can you confirm?

image

**Comment by msteiger** [comment]
Scaling the font artificially by 20% gives me this:

image

Looks good to me. What do you think?

**Comment by mdj117** [comment]
Both look good enough

**Comment by msteiger** [comment]
Apparently, Nashk also uses composite characters, which are not created / rendered correctly. Maybe [Hiero](https://github.com/libgdx/libgdx/wiki/Hiero) can help us here?

**Comment by mdj117** [comment]
Could any of the admins merge this PR ?

**Comment by mdj117** [comment]
@gooey ok to test

**Comment by mdj117** [comment]
@gooeyhub: ok to test

**Comment by rzats** [comment]
@mdj117: Gooey's tests can only be activated by admins. Either way, none of the issues from my initial review were addressed, so this isn't ready to be merged just yet.

**Comment by mdj117** [comment]
Couldn't find the words you were looking for :(
I can do a simple google translate if that works

File: engine/src/main/resources/assets/i18n/menu_fa.lang (hunk 10 above)
Comment (mdj117) [inline engine/src/main/resources/assets/i18n/menu_fa.lang]: بزرگنمایی

File: engine/src/main/resources/assets/i18n/menu_fa.lang (hunk 3 above)
Comment (VaibhavBajaj) [inline engine/src/main/resources/assets/i18n/menu_fa.lang]: @mdj117 Try "Key already does an action. Do you want to change it anyway?"
The underlying meaning is what counts, not the exact words.

File: engine/src/main/resources/assets/i18n/menu_fa.lang (hunk 1 above)
Comment (VaibhavBajaj) [inline engine/src/main/resources/assets/i18n/menu_fa.lang]: @mdj117 In Hindi, I just wrote NUI in Hindi such that the syllables pronunciation sounded the same. For editor, I am sure there would be an Arabic word meaning "Editor", "Changer" or "Property Control".
The underlying meaning is what counts, not the exact words.

**Comment by msteiger** [comment]
The most important issue I see here is that the Nashk glyphs are not rendered correctly by BMFont. Also, Hiero fails us here. I haven't found any good resources on how to do it right. I'll try to look into this, but I can be tough.

Feel free to investigate yourself on how to render the text. Similar discussion is here: https://github.com/MovingBlocks/Terasology/pull/2733

**Comment by Cervator** [comment]
Merged this even if incomplete as part of trying to clean up all the things :-)

Also updated Weblate and this shows up there as "Persian" nearly complete.

@mdj117: if you want to double check that Weblate looks correct please do! We'll get the last few things translated eventually. Thanks!

In-game the language is now excluded as part of #2784 until we can get the font set up and working

Fun side-note: The .lang file from here prompted IntelliJ for how to open it despite me having another .lang file open already. And it edits right-to-left since Farsi is supposed to work that way. Must be the encoding and IntelliJ being smart. Yet that also meant the symbols showed in reverse direction between the GitHub diff here and the file in my IntelliJ :D

**Pull request merged**